README.md
# NAME

Acme::MadokaMagica - It's miracles and magic are real module.

# SYNOPSIS

    use Acme::MadokaMagica;

    my($madoka,$homura,$mami,$kyouko,$sayaka) = Acme::MadokaMagica->main_members;

    print $madoka->name;         # '鹿目 まどか';
    print $madoka->firstname;    # 'まどか';
    print $madoka->lastname;     # '鹿目';
    print $madoka->age;          #  14;
    print  $madoka->birthday;    #  '10/3';
    print  $madoka->blood_type;  #  'A';
    print  $madoka->cv;          # '悠木碧';
    print  $madoka->say;         #   'ウェヒヒww';
    print  $madoka->color;       #   'Pink';
    print  $madoka->qb;
    print  $madoka->name;        # 'Kriemhild_Gretchen';
    print  $madoka->color;       #   'black';

    my ($mami) = Acme::MadokaMagica->alone_members;
    print $mami->say; #ティロ・フィナーレ

    my ($kyoko,$sayaka) = Acme::MadokaMagica->members_of($Acme::MadokaMagica::KyoSaya);

    print $kyoko->say; #'喰うかい?';
    $sayaka->qb;
    print $sayaka->name; #'Oktavia_Von_Seckendorff'


# DESCRIPTION

MadokaMagica is one of the most famouse Japanese TV animation.
This animation is magical girl heartful story.
When you leave 100 lines, the soul gem makes it impossible to use method.

It was in reference Acme::PriPara (C)htk291.

# LICENSE

Copyright (C) AnaTofuZ.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

AnaTofuZ <e155730@ie.u-ryukyu.ac.jp>


Changes
Revision history for Perl extension Acme-MadokaMagica

{{$NEXT}}

0.09 2019-02-01T09:28:39Z
    - remove debug print...

0.08 2019-02-01T09:25:53Z
    - not use Mouse and fix soul gem calc

0.07 2017-03-16T11:12:20Z
    - add a limit 100 line for soul gem

0.06 2017-03-11T01:00:18Z
    - use methods

0.05 2017-03-04T07:00:40Z
    -   modofy class name

0.04 2017-03-04T06:03:01Z
    - delete line arg

0.03 2017-03-04T05:52:14Z
    - fix hashref

0.02 2017-03-04T05:27:39Z
    - add cpanfile


0.01 2017-03-04T05:14:38Z

    - original version



lib/Acme/MadokaMagica/TvMembers.pm
package Acme::MadokaMagica::TvMembers;
use strict;
use warnings;
use utf8;
use Data::Section::Simple;
use YAML::Tiny;

sub new {
    my ($class, %args) = @_;

    my $self = { _has_qb => undef };
    my $ds = Data::Section::Simple->new($class);
    my $sections = $ds->get_data_section;
    for my $section_name ( keys %{$sections} ) {
        my $yml = YAML::Tiny->read_string($sections->{$section_name});
        my $member_info = $yml->[0];
        for my $key ( keys %{$member_info} ) {
            $self->{$key} = $member_info->{$key};
        }
    }
    if (defined $args{line}) {
        $self->{startline} = $args{line};
    } else {
        $self->{startline} = (caller)[2];
    }

    return bless $self, $class;
}

sub has_qb {
    my $self = shift;

    if (@_){
      $self->{_has_qb} = shift;
    }

    return $self->{_has_qb};
}

sub name {
    my ($self) = @_;
    my $line = (caller)[2];

    my $limit = $self->{startline} +100;
    print "$line : $limit\n";
    if( $line >= $limit ) {
        return undef;
    }
    return $self->has_qb ? $self->{witchename} : $self->lastname.' '.$self->firstname;
}

sub firstname {
    my $self = shift;
    my $line = (caller)[2];

    my $limit = $self->{startline} +100;
    if($line >= $limit ) {
        return undef;
    }
    return $self->{firstname};
}

sub birthday {
    my $self = shift;
    my $line = (caller)[2];

    my $limit = $self->{startline} +100;
    if($line >= $limit ) {
        return undef;
    }
    return $self->{birthday};
}

sub blood_type {
    my $self = shift;
    my $line = (caller)[2];

    my $limit = $self->{startline} +100;
    if( $line >= $limit ) {
        return undef;
    }
    return $self->{blood_type};
}

sub lastname {
    my $self = shift;
    my $line = (caller)[2];

    my $limit = $self->{startline} +100;
    if( $line >= $limit ) {
        return undef;
    }
    return $self->{lastname};
}

sub age {
    my $self = shift;
    my $line = (caller)[2];

    my $limit = $self->{startline} +100;
    if( $line >= $limit ) {
        return undef;
    }
    return $self->{age};
}

sub color{
    my ($self) = @_;
    my $line = (caller)[2];

    my $limit = $self->{startline} +100;
    if( $line >= $limit ) {
        return undef;
    }

    return $self->has_qb ? "black" : $self->{color};
}

sub qb {
    my ($self) = @_;
    my $line = (caller)[2];

    my $limit = $self->{startline} +100;
    if( $line >= $limit ) {
        return undef;
    }

    $self->has_qb(1);
}

sub say {
    my ($self) = @_;
    my $line = (caller)[2];

    my $limit = $self->{startline} +100;
    if( $line >= $limit ) {
        return undef;
    }
    return $self->{say};
}

sub cv {
    my $self = shift;
    my $line = (caller)[2];

    my $limit = $self->{startline} +100;
    if( $line >= $limit ) {
        return undef;
    }
    return $self->{cv};
}


1;


lib/Acme/MadokaMagica.pm
package Acme::MadokaMagica;
use 5.008001;
use strict;
use warnings;
use utf8;
use parent 'Exporter';

our $VERSION = "0.08";
our $miracle = "奇跡";
our $magical = "魔法";

our @EXPORT = qw[ $miracle $magical];

use Readonly;

Readonly our $KyoSaya => [
    "SakuraKyoko",
    "MikiSayaka",
];

Readonly our $MadoHomu => [
    "KanameMadoka",
    "AkemiHomura",
];

Readonly our $Alone => [
    "TomoeMami",
];

Readonly our $HollyQuintet => [
    @$MadoHomu,
    @$Alone,
    @$KyoSaya,
];

Readonly our $AloneMembers => [
    @$Alone,
];

Readonly our $MainMembers => [
    @$HollyQuintet,
];

sub alone_members {
    my $self = shift;
    return $self->members_of($AloneMembers,(caller)[2]);
}

sub main_members {
    my $self = shift;
    return $self->members_of($MainMembers,(caller)[2]);
}

sub members_of {
    my ($self, $team) = @_;
    my $line = $_[2] ||  (caller)[2];
    my @members;

    for my $member_name (@{ $team }){
        my $pkg = "Acme::MadokaMagica::TvMembers::$member_name";
        if (eval "require $pkg;1;"){
            push @members,$pkg->new("line" => $line);
        }
    }

    return @members;
}

1;
__END__

=encoding utf-8

=head1 NAME

Acme::MadokaMagica - It's miracles and magic are real module.

=head1 SYNOPSIS

  use Acme::MadokaMagica;

    my($madoka,$homura,$mami,$kyouko,$sayaka) = Acme::MadokaMagica->main_members;

    print $madoka->name;         # '鹿目 まどか';
    print $madoka->firstname;    # 'まどか';
    print $madoka->lastname;     # '鹿目';
    print $madoka->age;          #  14;
    print  $madoka->birthday;    #  '10/3';
    print  $madoka->blood_type;  #  'A';
    print  $madoka->cv;          # '悠木碧';
    print  $madoka->say;         #   'ウェヒヒww';
    print  $madoka->color;       #   'Pink';
    print  $madoka->qb;
    print  $madoka->name;        # 'Kriemhild_Gretchen';
    print  $madoka->color;       #   'black';

    my ($mami) = Acme::MadokaMagica->alone_members;
    print $mami->say; #ティロ・フィナーレ

    my ($kyoko,$sayaka) = Acme::MadokaMagica->members_of($Acme::MadokaMagica::KyoSaya);

    print $kyoko->say; #'喰うかい?';
    $sayaka->qb;
    print $sayaka->name; #'Oktavia_Von_Seckendorff'


=head1 DESCRIPTION

MadokaMagica is one of the most famouse Japanese TV animation.
This animation is magical girl heartful story.
When you leave 100 lines, the soul gem makes it impossible to use method.

It was in reference Acme::PriPara (C)htk291.

=head1 LICENSE

Copyright (C) AnaTofuZ.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

AnaTofuZ E<lt>e155730@ie.u-ryukyu.ac.jpE<gt>

=cut

